File: mobile/lib/relay_failover.dart
class RelaySelector {
  RelaySelector(Iterable<String> urls)
      : _urls = urls
            .map((url) => url.trim())
            .where((url) => url.isNotEmpty)
            .toList();

  final List<String> _urls;
  int _index = 0;

  String get current => _urls.isEmpty ? '' : _urls[_index % _urls.length];

  void connected() {}

  void reset() {
    _index = 0;
  }

  void failed() {
    if (_urls.isNotEmpty) {
      _index = (_index + 1) % _urls.length;
    }
  }
}

File: pc/clipsync/relay_failover.py
"""Sticky primary/backup selection for ClipSync relay connections."""

from __future__ import annotations


class RelaySelector:
    def __init__(self, urls: list[str] | tuple[str, ...]) -> None:
        self.urls = [url.strip() for url in urls if url and url.strip()]
        self.index = 0

    @property
    def current(self) -> str:
        if not self.urls:
            return ""
        return self.urls[self.index % len(self.urls)]

    def connected(self) -> None:
        """Keep the successful relay active until the socket fails."""

    def failed(self) -> None:
        if self.urls:
            self.index = (self.index + 1) % len(self.urls)
